File: Embedding-Application/backend/app/Support/DocumentParsing/CsvParser.php
<?php

namespace App\Support\DocumentParsing;

use Illuminate\Support\Str;
use RuntimeException;

class CsvParser implements DocumentParserInterface
{
    use TextNormalizer;

    public function supports(string $mimeType): bool
    {
        return in_array($mimeType, [
            'text/csv',
            'application/csv',
            'application/vnd.ms-excel',
        ]);
    }

    public function parse(string $path, array $options = []): ParsedDocument
    {
        $handle = fopen($path, 'r');
        if (!$handle) {
            throw new RuntimeException("Kan bestand niet openen: {$path}");
        }

        $delimiter = $options['delimiter'] ?? $this->detectDelimiter($path);
        $hasHeader = $options['has_header'] ?? true;
        $mapping = $options['mapping'] ?? null;

        $headers = [];
        $rows = [];
        $lineNumber = 0;

        while (($data = fgetcsv($handle, 0, $delimiter)) !== false) {
            $lineNumber++;

            if ($lineNumber === 1 && $hasHeader) {
                $headers = $data;
                continue;
            }

            $rows[] = $hasHeader ? array_combine($headers, $data) : $data;
        }

        fclose($handle);

        $docId = $options['doc_id'] ?? Str::slug(pathinfo($path, PATHINFO_FILENAME));
        $title = $options['title'] ?? pathinfo($path, PATHINFO_FILENAME);

        // Convert rows to sections based on mapping
        $sections = $this->rowsToSections($rows, $headers, $mapping, $docId);

        $metadata = [
            'row_count' => count($rows),
            'column_count' => count($headers),
            'columns' => $headers,
            'delimiter' => $delimiter,
        ];

        return new ParsedDocument(
            docId: $docId,
            title: $title,
            sections: $sections,
            metadata: $metadata,
            description: $options['description'] ?? null,
            category: $options['category'] ?? null,
        );
    }

    public function getMappableFields(string $path): array
    {
        $handle = fopen($path, 'r');
        if (!$handle) {
            return [];
        }

        $delimiter = $this->detectDelimiter($path);
        $headers = fgetcsv($handle, 0, $delimiter);
        fclose($handle);

        if (!$headers) {
            return [];
        }

        $fields = [];
        foreach ($headers as $index => $header) {
            $key = Str::slug($header) ?: "column_{$index}";
            $fields[$key] = $header;
        }

        return $fields;
    }

    public function requiresMapping(): bool
    {
        return true;
    }

    /**
     * Detect the CSV delimiter.
     */
    private function detectDelimiter(string $path): string
    {
        $handle = fopen($path, 'r');
        $line = fgets($handle);
        fclose($handle);

        if (!$line) {
            return ',';
        }

        $delimiters = [
            ',' => substr_count($line, ','),
            ';' => substr_count($line, ';'),
            "\t" => substr_count($line, "\t"),
            '|' => substr_count($line, '|'),
        ];

        return array_search(max($delimiters), $delimiters);
    }

    /**
     * Convert CSV rows to document sections.
     */
    private function rowsToSections(array $rows, array $headers, ?array $mapping, string $docId): array
    {
        // If no mapping provided, create one section with all data as text
        if (!$mapping) {
            $text = $this->rowsToText($rows, $headers);
            return [
                [
                    'title' => 'Data',
                    'slug' => 'data',
                    'order_index' => 0,
                    'text' => $this->normalizeText($text),
                    'metadata' => ['row_count' => count($rows)],
                ],
            ];
        }

        $sections = [];
        $titleColumn = $mapping['title_column'] ?? null;
        $contentColumn = $mapping['content_column'] ?? null;
        $groupColumn = $mapping['group_column'] ?? null;

        // Group by column if specified
        if ($groupColumn && isset($rows[0][$groupColumn])) {
            $grouped = [];
            foreach ($rows as $row) {
                $group = $row[$groupColumn] ?? 'Overig';
                $grouped[$group][] = $row;
            }

            $index = 0;
            foreach ($grouped as $groupName => $groupRows) {
                $text = '';
                foreach ($groupRows as $row) {
                    if ($titleColumn && isset($row[$titleColumn])) {
                        $text .= "### {$row[$titleColumn]}\n";
                    }
                    if ($contentColumn && isset($row[$contentColumn])) {
                        $text .= $row[$contentColumn] . "\n\n";
                    } else {
                        $text .= $this->rowToText($row, $headers) . "\n\n";
                    }
                }

                $sections[] = [
                    'title' => $groupName,
                    'slug' => Str::slug($groupName) ?: "groep-{$index}",
                    'order_index' => $index++,
                    'text' => $this->normalizeText($text),
                    'metadata' => ['row_count' => count($groupRows)],
                ];
            }
        } else {
            // Create one section per row if title column is specified
            if ($titleColumn) {
                foreach ($rows as $index => $row) {
                    $rowTitle = $row[$titleColumn] ?? "Rij " . ($index + 1);
                    $content = $contentColumn && isset($row[$contentColumn])
                        ? $row[$contentColumn]
                        : $this->rowToText($row, $headers);

                    $sections[] = [
                        'title' => $rowTitle,
                        'slug' => Str::slug($rowTitle) ?: "rij-{$index}",
                        'order_index' => $index,
                        'text' => $this->normalizeText($content),
                        'metadata' => ['row_data' => $row],
                    ];
                }
            } else {
                // Single section with all data
                $text = $this->rowsToText($rows, $headers);
                $sections[] = [
                    'title' => 'Data',
                    'slug' => 'data',
                    'order_index' => 0,
                    'text' => $this->normalizeText($text),
                    'metadata' => ['row_count' => count($rows)],
                ];
            }
        }

        return $sections;
    }

    /**
     * Convert all rows to readable text.
     */
    private function rowsToText(array $rows, array $headers): string
    {
        $text = '';
        foreach ($rows as $index => $row) {
            $text .= "--- Rij " . ($index + 1) . " ---\n";
            $text .= $this->rowToText($row, $headers) . "\n\n";
        }
        return trim($text);
    }

    /**
     * Convert a single row to readable text.
     */
    private function rowToText(array $row, array $headers): string
    {
        $lines = [];
        foreach ($row as $key => $value) {
            if (!empty($value)) {
                $label = is_numeric($key) && isset($headers[$key]) ? $headers[$key] : $key;
                $lines[] = "{$label}: {$value}";
            }
        }
        return implode("\n", $lines);
    }
}

File: app/backend/embedder/parsing/csv_parser.py
"""CSV parser met optionele kolom-mapping.

Equivalent van de Laravel ``CsvParser``. Vereist een ``mapping``-stap (vandaar
``requires_mapping() -> True``); zonder mapping gaat alle data in één ``Data``-
sectie.
"""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any, Optional

from .normalizer import normalize_text
from .parsed import ParsedDocument, slugify


_SUPPORTED_DELIMITERS = (",", ";", "\t", "|")


class CsvParser:
    _SUPPORTED_MIMES = {
        "text/csv",
        "application/csv",
        "application/vnd.ms-excel",
    }

    def supports(self, mime_type: str) -> bool:
        return mime_type in self._SUPPORTED_MIMES

    def parse(self, path: str, options: dict[str, Any]) -> ParsedDocument:
        file_path = Path(path)
        delimiter = options.get("delimiter") or self._detect_delimiter(path)
        has_header = options.get("has_header", True)
        mapping = options.get("mapping")

        headers: list[str] = []
        rows: list[dict] = []

        with file_path.open("r", encoding="utf-8-sig", newline="") as handle:
            reader = csv.reader(handle, delimiter=delimiter)
            for line_index, data in enumerate(reader):
                if line_index == 0 and has_header:
                    headers = [str(cell) for cell in data]
                    continue
                if has_header and headers:
                    row = {
                        headers[i] if i < len(headers) else f"column_{i}": value
                        for i, value in enumerate(data)
                    }
                else:
                    row = {f"column_{i}": value for i, value in enumerate(data)}
                rows.append(row)

        doc_id = options.get("doc_id") or slugify(file_path.stem)
        title = options.get("title") or file_path.stem

        sections = self._rows_to_sections(rows, headers, mapping)

        metadata = {
            "row_count": len(rows),
            "column_count": len(headers),
            "columns": headers,
            "delimiter": delimiter,
        }

        return ParsedDocument(
            doc_id=doc_id,
            title=title,
            sections=sections,
            metadata=metadata,
            description=options.get("description"),
            category=options.get("category"),
        )

    def get_mappable_fields(self, path: str) -> dict[str, str]:
        file_path = Path(path)
        try:
            delimiter = self._detect_delimiter(path)
            with file_path.open("r", encoding="utf-8-sig", newline="") as handle:
                reader = csv.reader(handle, delimiter=delimiter)
                headers = next(reader, None)
        except OSError:
            return {}
        if not headers:
            return {}
        fields: dict[str, str] = {}
        for index, header in enumerate(headers):
            key = slugify(header, fallback=f"column_{index}")
            fields[key] = header
        return fields

    def requires_mapping(self) -> bool:
        return True

    # ---- internals ---------------------------------------------------------

    def _detect_delimiter(self, path: str) -> str:
        try:
            with open(path, "r", encoding="utf-8-sig") as handle:
                line = handle.readline()
        except OSError:
            return ","
        if not line:
            return ","
        counts = {delim: line.count(delim) for delim in _SUPPORTED_DELIMITERS}
        return max(counts, key=counts.get) or ","

    def _rows_to_sections(
        self,
        rows: list[dict],
        headers: list[str],
        mapping: Optional[dict],
    ) -> list[dict]:
        if not mapping:
            text = self._rows_to_text(rows, headers)
            return [
                {
                    "title": "Data",
                    "slug": "data",
                    "order_index": 0,
                    "text": normalize_text(text),
                    "metadata": {"row_count": len(rows)},
                }
            ]

        title_column = mapping.get("title_column")
        content_column = mapping.get("content_column")
        group_column = mapping.get("group_column")

        if group_column and rows and group_column in rows[0]:
            grouped: dict[str, list[dict]] = {}
            for row in rows:
                key = row.get(group_column) or "Overig"
                grouped.setdefault(str(key), []).append(row)

            sections: list[dict] = []
            for index, (group_name, group_rows) in enumerate(grouped.items()):
                parts: list[str] = []
                for row in group_rows:
                    if title_column and title_column in row:
                        parts.append(f"### {row[title_column]}")
                    if content_column and content_column in row:
                        parts.append(str(row[content_column]))
                        parts.append("")
                    else:
                        parts.append(self._row_to_text(row, headers))
                        parts.append("")
                sections.append(
                    {
                        "title": group_name,
                        "slug": slugify(group_name, fallback=f"groep-{index}"),
                        "order_index": index,
                        "text": normalize_text("\n".join(parts)),
                        "metadata": {"row_count": len(group_rows)},
                    }
                )
            return sections

        if title_column:
            sections = []
            for index, row in enumerate(rows):
                row_title = row.get(title_column) or f"Rij {index + 1}"
                content = (
                    str(row[content_column])
                    if content_column and content_column in row
                    else self._row_to_text(row, headers)
                )
                sections.append(
                    {
                        "title": str(row_title),
                        "slug": slugify(str(row_title), fallback=f"rij-{index}"),
                        "order_index": index,
                        "text": normalize_text(content),
                        "metadata": {"row_data": row},
                    }
                )
            return sections

        text = self._rows_to_text(rows, headers)
        return [
            {
                "title": "Data",
                "slug": "data",
                "order_index": 0,
                "text": normalize_text(text),
                "metadata": {"row_count": len(rows)},
            }
        ]

    def _rows_to_text(self, rows: list[dict], headers: list[str]) -> str:
        parts: list[str] = []
        for index, row in enumerate(rows):
            parts.append(f"--- Rij {index + 1} ---")
            parts.append(self._row_to_text(row, headers))
            parts.append("")
        return "\n".join(parts).strip()

    @staticmethod
    def _row_to_text(row: dict, headers: list[str]) -> str:
        lines = []
        for key, value in row.items():
            if value in (None, ""):
                continue
            lines.append(f"{key}: {value}")
        return "\n".join(lines)
